Implement DPP damage formula

Nineage opened the pull request:
Inspired by the chain of events in this thread: http://www.smogon.com/forums/threads/bugs-that-directly-influence-the-outcome-of-a-game.3603259/#post-7347664

Basically, while later generations implement most of their modifiers on damage (such as boosts by items, abilities, etc) in one place in the damage calculation, Gen 4 does this in several steps. This creates small discrepancies between expected damage output and PS damage at Level 100, and these become more significant as levels decrease, becoming potentially game-determining at Level 5.

I based the code on http://www.smogon.com/dp/articles/damage_formula which is what the PS damage calc is based on. I made the changes within the context of PS's event system, basically just adding clones of ModifyDamage (ModifyDamageOne and ModifyDamageTwo) for modifiers at different stages in the calculation. The third modifier is still named ModifyDamage so that if modifiers that didn't exist in Gen 4 appear (in a custom game maybe?) they will still have an effect on the damage rolls.

In every test I've done, the damage has been equal to both what I've calculated by hand and what the damage calculator shows. Obviously, that doesn't mean the code is perfect, and because this is such an important function, I will work to implement any and all feedback ASAP.

Tagging @Marty-D because I think he does mechanics stuff.

Zarel commented:
Entirely replacing the damage formula isn't really a great idea, unfortunately. I was hoping to get an overview of every generation's damage formula so I could do it in a more subtle way...

Nineage commented:
@Zarel maybe a potential option would be splitting the ModifyDamage event in the most recent damage calculator and then calling all 3 modifiers consecutively if (gen > 4)? The issue is that certain things like weather modifiers occur in such a different place in gen 4 that changing the modern calculator would lead to a lot of repeated code at different points in the same function. 

Is there a specific reason this implementation doesn't work? Is it performance-based or more of a preference thing in how these types of things are implemented?

Zarel commented:
It's just replacing a huge function with another huge function. I'd rather split the function up into smaller functions so they could be replaced piecemeal, I think?

TheFenderStory commented:
I'm not sure how practical it would be to piecemeal out these damage calculation functions. My understanding is that they are pretty unique between different generations.

Zarel commented:
The DPP and BW damage formulas appear identical, other than a few modifiers being in different places.

Nineage commented:
@Zarel I believe I've implemented what we talked about both here and on Discord. Essentially, all of the changed components except burn damage lie between the initial basePower calculation and the randomizer, so I exported those into a `getInitialMods` function. I also changed the names of the modification stages to ModifyDamagePhaseX.

I'm still unsure if this is the optimal place to splice the getDamage function, but I think this is a good play to keep it modular without sacrificing readability.

Zarel commented on sim/battle.js:
```diff
@@ -1873,6 +1861,26 @@ class Battle extends Dex.ModdedDex {
 	randomizer(baseDamage) {
 		return Math.floor(baseDamage * (100 - this.random(16)) / 100);
 	}
+	getInitialMods(pokemon, target, move, baseDamage) {
```
Hm, I'm not sure if this is the point at which I'd split up the two functions.

Mostly, it's probably all the complicated code to calculate Attack and Defense which I feel like should be in a separate function from the rest of the `getDamage` code. One idea might be to put the entire rest of it in a `calculateDamage` function. Or maybe just `modifyDamage` for the entire calculation after `baseDamage`.

Nineage commented:
@Zarel I chose the second option because the first would've required passing even more variables between functions I think, and this one makes more sense to me in terms of readability (damage is set up, then modified, then returned). The one thing I really hate is passing both basePower and baseDamage between functions but they are both necessary in both, and I don't see an easy way to calculate it twice without annoying duplicate code.

Zarel commented on sim/battle.js:
```diff
@@ -1795,7 +1793,18 @@ class Battle extends Dex.ModdedDex {
+		let baseDamage = Math.floor(Math.floor(Math.floor(2 * level / 5 + 2) * basePower * attack / defense) / 50);
+		
+		// Calculate initial damage modifiers (order differs between generations)
+		baseDamage = this.modifyDamage(pokemon, target, move, basePower, baseDamage, suppressMessages);
```
Apparently, `basePower` is passed because it can be zero sometimes. In what situations is this true?

Nineage replied:
Hm...I passed this because it was in the original code so I figured there was something I was missing, but I've gone through each event and there's nothing that can modify basePower after ``if (!basePower) return 0;`` so I guess it's fine to assume it isn't 0. Makes things prettier anyway. Code has been updated.

Zarel commented on sim/battle.js:
```diff
@@ -1864,11 +1873,11 @@ class Battle extends Dex.ModdedDex {
 		}
 
-		return Math.floor(baseDamage);
+		return baseDamage;
```
I think this floor should be kept.

Zarel commented on sim/battle.js:
```diff
@@ -1795,7 +1793,18 @@ class Battle extends Dex.ModdedDex {
+		// Calculate initial damage modifiers (order differs between generations)
+		baseDamage = this.modifyDamage(pokemon, target, move, baseDamage, suppressMessages);
+
+		return Math.floor(baseDamage);
```
And then you could just `return this.modifyDamage(...)`

Zarel commented on sim/battle.js:
```diff
@@ -1795,7 +1793,18 @@ class Battle extends Dex.ModdedDex {
+
+		return Math.floor(baseDamage);
+	}
+	modifyDamage(pokemon, target, move, baseDamage, suppressMessages) {
```
Minor nitpick, but `baseDamage` should be the first argument, since that's what's being modified.

Nineage commented:
Updated. 
As a side note, I've noticed this is actually a great place to split getDamage for Gen III calc as well, so +1 for this approach I guess.

Zarel commented on sim/battle.js:
```diff
@@ -1795,7 +1793,16 @@ class Battle extends Dex.ModdedDex {
+	modifyDamage(baseDamage, pokemon, target, move, suppressMessages) {
+		if (!move.type) move.type = '???';
+		let type = move.type;
+		
```
Trailing spaces...

Zarel commented:
Okay, I've reviewed everything else and it looks good! We can merge this after the trailing space fix.

Nineage commented:
Removed the trailing space and double checked eslint, this should be good to go

_Sent from my Samsung SM-G920V using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_
